### ISSUE TITLE
Investigate using parallel IO

### ISSUE THREAD

**Issue by aekiss** [COSIMA/cice5#34]
It may be worth trying to compile with parallel IO using [PIO](https://ncar.github.io/ParallelIO/) (`setenv IO_TYPE  pio`).

We currently compile CICE with serial IO (`setenv IO_TYPE  netcdf` in `bld/build.sh`), so one CPU does all the IO and we end up with an Amdahl's law situation that limits the scalability with large core counts.

At 0.1 deg CICE is IO-bound when doing daily outputs (see `Timer  12` in `ice_diag.d`), and the time spent in CICE IO accounts for almost all the time MOM waits for CICE (`oasis_recv` in `access-om2.out`) so the whole coupled model is waiting on one cpu. With daily CICE output at 0.1deg this is ~19% of the model runtime (it's only ~2% without daily CICE output). Lowering the compression level to 1 (https://github.com/COSIMA/cice5/issues/33) has helped (MOM wait was 23% with level 5), and omitting static field output (https://github.com/COSIMA/cice5/issues/32) would also help.

Also I understand that PIO doesn't support compression - is that correct?

@russfiedler had these comments on [Slack](https://arccss.slack.com/archives/C9Q7Y1400/p1557272377089800):
> I have a feeling that the CICE parallel IO hadn't really been tested or there was some problem with it.
> We would have to update the netcdf versions being used in CICE for a start.
> the distributors of PIO note that they need to use netCDF 4.6.1 and HDF5 1.10.4 or later for their latest version. There's a bug in parallel collective IO in earlier hdf5 versions. The NCI version of netCDF 4.6.1 is built with hdf5 1.10.2! marshall noted above that Rui found a performance drop off when moving from 1.10.2 to 1.10.4.
> the gather is done on all the small tiles. So you have each PE sending a single  horizontal slab several times to the root PE for each level.
> the number of MPI calls is probably the main issue. It looks like there's an individual send/recv for each tile rather than either a bulk send of the tiles or something more funky using MPI_Gather(v) and MPI_Type_create_subarray.

Slack discussion: https://arccss.slack.com/archives/C9Q7Y1400/p1557272377089800

**Comment by nichannah** [comment]
I've been looking at the CICE PIO code. It is not as complete as the serial netcdf code, for example it doesn't do error proper checking. The PIO code still exists and is documented in CICE6. 

My next step is to try it out on a 1deg case. 

Another option, which may be better even if PIO works is to take the MOM5 approach and have each PE output to it's own file followed by an offline collate. The advantage of this would be that we can continue to use the existing netcdf code (with slight modifications). The down-side would be that we need to write a collate program.

**Comment by russfiedler** [comment]
@nichannah  I think with the moves by Ed Hartnett wrt implementing PIO in FMS I think it would be best to go the PIO route to stay reasonably compatible with future FMS and CICEn.

**Comment by nichannah** [comment]
Steps to build PIO cice.

1. download and extract pio:

```
wget https://github.com/NCAR/ParallelIO/releases/download/pio2_4_3/pio-2.4.3.tar.gz
tar zxvf pio-2.4.3.tar.gz
```

2. load necessary modules:

```
module load intel-fc/17.0.1.132
module load intel-cc/17.0.1.132
module load openmpi/3.0.1
module load netcdf/4.6.1p
module load parallel-netcdf/1.6.1
```

I also tried openmpi/1.10.2 but the build failed with link errors.

3. set environment variables

```
export CPPFLAGS='-std=c99 -I${NETCDF}/include/ -L${PARALLEL_NETCDF_BASE}/include/'
export LDFLAGS='-L${NETCDF}/lib/ -L${PARALLEL_NETCDF_BASE}/lib/'
```

4. configure and make

```
./configure --enable-fortran --prefix=/short/x77/nah599/access-om2/src/cice5/pio-2.4.3/usr
make
```

**Comment by nichannah** [comment]
I looks like we need to grab some CESM code, the PIO code makes use of something called shr_pio_mod. The code can be found here:

https://github.com/CESM-Development/cesm-git-experimental/tree/master/cesm/models/csm_share

**Comment by nichannah** [comment]
Followed my instructions as above with new versions and the configure step hangs. Assuming this is a transient raijin system problem so leaving for now.

**Comment by nichannah** [comment]
Current status is that PIO is building, need to modify CICE PIO support so that it works without CESM dependencies.

**Comment by nichannah** [comment]
The PIO code is ready to be tested however there is a problem with netcdf, compiler and openmpi version compatibility between the new CICE and the rest of the model. So this issue is now dependent on upgrading these things.

**Comment by aekiss** [comment]
In ACCESS-OM2 sea ice concentration is passed to MOM via OASIS
https://github.com/COSIMA/01deg_jra55_iaf/blob/30df8f5fd6404aeb459ff44298936df576dfbbf0/namcouple#L295
so we could output that field in parallel via MOM.

I couldn't find a relevant diagnostic here
https://github.com/COSIMA/access-om2/wiki/Technical-documentation#MOM5-diagnostics-list
so it looks like we'd need to write one.

**Comment by russfiedler** [comment]
I've put this in the WOMBAT version but I've been holding off on issueing a pull request until @nichannah updates the way he proposes to pass new fields.

https://github.com/russfiedler/MOM5/blob/wombat/src/mom5/ocean_core/ocean_sbc.F90#L5971

**Comment by russfiedler** [comment]
Also as a note to above. netCDF on gadi should be suitable for PIO

**Comment by nichannah** [comment]
Updated PIO build instructions: 

```
cd $ACCESS_OM_DIR/src/cice5
wget https://github.com/NCAR/ParallelIO/releases/download/pio2_5_0/pio-2.5.0.tar.gz
tar zxvf pio-2.5.0.tar.gz
cd pio-2.5.0
module load intel-compiler/2019.5.281
module load netcdf/4.7.1p
module load openmpi/4.0.2
export CPPFLAGS=-I${NETCDF}/include/
export LDFLAGS=-L${NETCDF}/lib/ompi3/
./configure --enable-fortran --disable-pnetcdf --prefix=$ACCESS_OM_DIR/src/cice5/pio-2.5.0/usr
make
make install
```

**Comment by nichannah** [comment]
preliminary results from a 10 day 0.1 run with daily cice output. previously writing output was 15% of CICE runtime, it’s now 6%.
10:01
mom now spending less than half as much time waiting on ice. from 12% or runtime down to 5%

**Comment by aekiss** [comment]
Thanks @nichannah, that's great news.

Did you run your test with 799 CICE cores? And am I right in thinking CICE with PIO uses all cores (rather than a subset like MOM io_layout)? If so, I'm a little surprised it didn't speed up more, if there are 799x more cores doing the output. I guess there's some extra overhead in PIO?

@marshallward's tests on Raijin showed CICE would scale well up to about 2000 cores and is still reasonable at 3000 (see table below). If so, I guess we'd need over 4000 CICE cores to get below 1% MOM wait time, which seems rather a lot. But in our standard configs (serial CICE io, monthly outputs) MOM spends just under 2% of its time waiting for CICE, so 1% is better than we're used to.

![Screen Shot 2020-05-03 at Sun 3-5 9 48am](https://user-images.githubusercontent.com/31054815/80895021-c8c5cf00-8d23-11ea-9c27-93ca8595e118.png)

**Comment by nichannah** [comment]
Thanks @aekiss, that's useful. 

I'm now running a test to see how a run with daily output compares to one with monthly output. If that is OK then perhaps we can start to use this feature before spending more time on optimisation.

**Comment by russfiedler** [comment]
I believe PIO allows some sort of flexibilty with which PEs are used https://ncar.github.io/ParallelIO/group___p_i_o__init.html . I don't know how flexible this is in what has been written for CICE. There is an interesting point made in the FAQ that it's sometimes worth moving the IO away from the root PE/task (and I presume node) due to the heavier load there.
Would it be worth investigating striping the files?

**Comment by nichannah** [comment]
Yes, it looks like there's some configuration optimisation that we can do with this. Presently I'm just using the simplest config which is a stride of 1 - so all procs are writing output. 

I have just completed two 2 month runs: 

1) standard config with mostly monthly cice output (16Gb output over 2 months)
2) PIO config with all daily output (460Gb output over 2 months)

Basically 1) is doing about 8Gb per month and 2) is doing 8Gb per day. 

The runtime of these two runs is almost identical. Looking at ice_diag.d the time taken for writing out history is similar but the PIO case is about 5% slower. See

```
/scratch/v45/nah599/access-om2/archive/01deg_jra55_iaf/output000/ice/ice_diag.d
/scratch/v45/nah599/access-om2/archive/pio_01deg_jra55_iaf/output000/ice/ice_diag.d
```

Incidentally, there seems to be something strange happening with the atm halo timers in the new PIO run. The mean time in the PIO run is 6 seconds but for the regular run it is 106 seconds. A possible explanation for this is that the PEs within CICE are better matched so collective operations don't have to wait as long on lagging PEs. 

So this new feature should allow daily ice output with no performance penalty over the existing configuration. I think it makes sense to merge this into master. Any objections? @aekiss?

Future work will involve looking at the scaling and performance of the whole model in more detail and at that point I can look at the different configuration options of PIO if ice output is a bottleneck.

**Comment by aekiss** [comment]
That's great that daily output can be done with nearly the same runtime. If you're confident that the output with PIO is bitwise identical to the non-PIO version then I see no reason not to merge into master, given that it makes daily output practical.
@AndyHoggANU any objections?

Also is compressed output still possible with PIO?

**Comment by AndyHoggANU** [comment]
Yes, I would like to see PIO included if at all possible. It would make it feasible to put *some* daily ice fields out in the IAF run, which would be a big benefit.

**Comment by aekiss** [comment]
According to timer 12, PIO with daily output is slightly faster than serial with monthly output (not 5% slower):
 
`/scratch/v45/nah599/access-om2/archive/01deg_jra55_iaf/output000/ice/ice_diag.d`
```
Timer   1:     Total   13783.11 seconds
Timer  12: ReadWrite     934.32 seconds
```

`/scratch/v45/nah599/access-om2/archive/pio_daily_01deg_jra55_iaf/output000/ice/ice_diag.d`
```
Timer   1:     Total   13728.74 seconds
Timer  12: ReadWrite     871.72 seconds
```

**Comment by nichannah** [comment]
Seems like a cleaner timer to look at. I was looking at History.

**Comment by russfiedler** [comment]
The big one is still the time that the ice model is waiting for MOM for coupling (Timer 18: 5800s). I think/hope that this now means that upping the core count for MOM should be efficient since that occasional slowdown for CICE output won't be occurring.

**Comment by nichannah** [comment]
Good point. Successful balancing between models depends a lot on the balance within a model.

**Comment by aekiss** [comment]
@nichannah your build instructions [above](https://github.com/COSIMA/cice5/issues/34#issuecomment-619974388) include a manual download of PIO - would this be better done as a PIO submodule within the CICE5 repo so that a recursive clone will get all the dependencies?

**Comment by aekiss** [comment]
also is your `34-pio` branch up to date on https://github.com/COSIMA/cice5 ? It only seems to have changes to `bld/config.nci.auscom.360x300` but not the other resolutions
https://github.com/COSIMA/cice5/compare/34-pio..master

**Comment by nichannah** [comment]
Yes I'll clean up the build process. A sub module within the cice repo is a good idea

**Comment by nichannah** [comment]
I don't think it is up to date. I'll need to do a bit more work to get things in a mergable state.

**Comment by aekiss** [comment]
no worries, let me know when you have a configuration that I can try building

**Comment by russfiedler** [comment]
Do we really want to support PIO as another submodule rather than pushing to get it supported on the system? It really has a much wider use and if we're not intending to do development on it I'm not sure that it should be part of the distribution.

**Comment by aekiss** [comment]
Fair point. I don't have an opinion either way, just so long as there's a seamless way to clone and build it on gadi.

**Comment by nichannah** [comment]
Current status: 

I've been doing further runs with the 0.1 deg and got a few segfaults at the end of the run. I've also noticed that there are several deficiencies in the way that the PIO interface is being used within CICE (the 32, 64 bit problem above, no error checking, bad ordering of setup and tear-down calls) so need to do a more thorough code review.

**Comment by aidanheerdegen** [comment]
> Do we really want to support PIO as another submodule rather than pushing to get it supported on the system? It really has a much wider use and if we're not intending to do development on it I'm not sure that it should be part of the distribution.

@rxy900 might be able to comment about how to go about getting this supported on the system.

**Comment by aekiss** [comment]
Nic has already put in a request to have it installed as a module

**Comment by nichannah** [comment]
Regarding the PIO module. Andrey has pointed out that it may not be worthwhile if ACCESS-OM2 is the only user.

Also I've found that it is quite useful to be able to recompile (e.g. with debug output) and perhaps this will be needed in the future if we do further optimisation/tuning work. 

I think it makes sense to compiler ourselves for the time being and then perhaps move to a central install down the track. Perhaps we could install it as a hh5 module?

**Comment by nichannah** [comment]
New best run is here:

/scratch/v45/nah599/access-om2/archive/pio_daily_rearr_box_01deg_jra55_iaf/output000/ice/ice_diag.d

History writing is down from around 900 secs for the plain netcdf monthly output to 300 secs for the PIO daily.

**Comment by aekiss** [comment]
Great work @nichannah - here are some results from a production test run compared to the same config without PIO. 
CICE I/O time reduced by 77%
Improved load balance: MOM wait for CICE reduced by 70% (from 12.1% to 3.7% of MOM total runtime)
overall SU cost and walltime reduced by 8.3%. We are now very close to 6mo/submit.

  | Control directory | Output directory | Job Id | Service Units | Walltime Used (hr) | Fraction of MOM runtime in oasis_recv | Max MOM wait for oasis_recv (s) | Max CICE wait for coupler (s) | Max CICE I/O time (s) | Memory Used (Gb) | NCPUs Used | MOM NCPUs | CICE NCPUs
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
serial IO | /home/156/aek156/payu/01deg_jra55v140_iaf_cycle2 | /scratch/x77/aek156/access-om2/archive/01deg_jra55v140_iaf_cycle2/output356 | 10897190 | 29119.68 | 2.80861111 | 0.121 | 1242.67672 | 2821.14 | 1242.19 | 3624.96 | 5184 | 4358 | 799
PIO | /home/156/aek156/payu/01deg_jra55v140_iaf_cycle2_pio_test | /scratch/x77/aek156/access-om2/archive/01deg_jra55v140_iaf_cycle2_pio_test/output356 | 10912826 | 26709.12 | 2.57611111 | 0.037 | 368.921135 | 2517.83 | 280.2 | 3788.8 | 5184 | 4358 | 799
change (%) |   |   | 0.14% | -8.28% | -8.28% | -69.42% | -70.31% | -10.75% | -77.44% | 4.52% | 0.00% | 0.00% | 0.00%

**Comment by aekiss** [comment]
Hi @nichannah I have 0.1deg PIO test run output in
`/scratch/x77/aek156/access-om2/archive/01deg_jra55v140_iaf_cycle2_pio_test/output356/ice/OUTPUT`
that should be identical to the non-PIO case here
`/scratch/x77/aek156/access-om2/archive/01deg_jra55v140_iaf_cycle2/output356/ice/OUTPUT` 
but isn't. I'm still getting to the bottom of it. 

Some differences are unimportant metadata differences, but there are also some cpu mask differences in grid variables, e.g. in `TLON` (PIO version is on the right):
![Screen Shot 2020-09-14 at Mon 14-9 9 18pm](https://user-images.githubusercontent.com/31054815/93079694-dfb3ff00-f6cf-11ea-8823-9cdda6183ce4.png)
These mask differences don't seem to be present in the data variables, only the grid ones.

**Comment by nichannah** [comment]
Thanks @aekiss for finding this. The blkmask variable also looks significantly different. I'll look into it as well, I have a couple of vague ideas about what might be causing this.

**Comment by nichannah** [comment]
Hi @aekiss, I have fixed this and pushed changes to my 01deg_jra55v140_iaf_cycle2_with_pio pull request. I needed to update the CICE executable. The problem was that I had modified the CICE block distribution algorithm to make sure each CPU had an equal number of blocks. This is not needed so in general I've changed it to a namelist option.

**Comment by aekiss** [comment]
ok thanks @nichannah , let me know when you've updated the exe and I'll give it another whirl

**Comment by nichannah** [comment]
@aekiss, the exe is here:

/g/data/ik11/inputs/access-om2/bin/cice_auscom_3600x2700_722p_7c74942_libaccessom2_d914095.exe

It should be referenced in the config.yaml that I pushed to the pull request on your branch.

**Comment by aekiss** [comment]
thanks @nic I have a test running from `/home/156/aek156/payu/01deg_jra55v140_iaf_cycle2_pio_test2`

**Comment by aekiss** [comment]
Looks like you've nailed it @nichannah - apart from some expected changes in a few global attributes the output from the test run is identical to the previous one according to this test:
https://github.com/aekiss/notebooks/blob/master/check_pio.ipynb
So as far as I can see it's ready for production use.

**Comment by aekiss** [comment]
reopening - I hadn't checked the restarts and the `iced.*` restarts seem to have the same processor mask issue, 
eg `Tsfcn` in
`/scratch/v14/pas548/restarts/KEEP/restart356/ice/iced.1986-04-01-00000.nc` (left) and
`/scratch/x77/aek156/access-om2/archive/01deg_jra55v140_iaf_cycle2_pio_test2/restart356/ice/iced.1986-04-01-00000.nc` (right)
![Screen Shot 2020-09-22 at Tue 22-9 1 29pm](https://user-images.githubusercontent.com/31054815/93841712-cf0b1680-fcd7-11ea-8818-b8596bff6ed7.png)

@nichannah can you have a look at this please?

**Comment by nichannah** [comment]
Hi @aekiss, 

The reason for the above is that the old approach was to:

1. master task fills a global sized array with a default value (in this case 0)
2. it then gathers restart field and puts them in above array
3. write out

So the land was explicitly written out at 0. Whereas in the new approach the PEs each write out the part of the domain that they cover and the rest is left as the netcdf fillvalue.

So I think that in this case the new way is actually not incorrect - values over land should be undefined rather than set to 0. 

The possible problem is that when we change number or layout of CICE CPUs the missing values may not be in exactly the same places, which makes it difficult to use restarts on an altered configuration. Much like the ocean model where it's necessary to collate when restarting with a changed config. 

Actually I think I've already hit this with one of my scaling configs - I'm gettting floating point exceptions and an ice thermo error as mentioned in the metting today.

**Comment by aekiss** [comment]
ah ok. So it it possible for the master task to also fill in the gaps? Or even simpler, set the fill value to 0?

**Comment by nichannah** [comment]
OK, so it doesn't look like CICE does a very thorough job of checking the land mask. My crash in: `ice_atmo.F90:352` does not have a check for the land mask. I think the solution to this is to fill land with 0 as before. The alternative is that we would need to do processing on restarts if we every need to use them for a new config.

**Comment by aekiss** [comment]
I agree that filling land with 0 seems the better option, rather than hoping we remember this gotcha into the indefinite future...

**Comment by nichannah** [comment]
The solution to this is not completely satisfactory. The obvious way to get netcdf to put 0's in places where no data is written is to set _FillValue = 0. This can be a bit confusing because there is no difference between "no data" and "data with value 0". However I think this is probably still better than the alternative which is needing to fix-up CICE restarts whenever the PE layout changes. 

See attached Tsfcn, the white has value 0 and the red is mostly -1.8. 

<img width="331" alt="Screen Shot 2020-10-08 at 10 49 09 pm" src="https://user-images.githubusercontent.com/921706/95454923-ce050500-09b8-11eb-80d0-121c6e432f42.png">

**Comment by aekiss** [comment]
I don't think setting _FillValue = 0 is a problem for this particular variable, as zero values are used for `Tsfcn` at land points outside the land mask in the PIO restarts anyway.

e.g. see `Tsfcn` in `/scratch/x77/aek156/access-om2/archive/01deg_jra55v140_iaf_cycle2_pio_test2/restart356/ice/iced.1986-04-01-00000.nc` (the range is narrowed for clarity):
![Screen Shot 2020-10-09 at Fri 9-10 4 36pm](https://user-images.githubusercontent.com/31054815/95547023-c0e62580-0a4d-11eb-8119-77113136ff59.png)

Setting _FillValue = 0 is consistent with what was done in the non-PIO restarts, which had zero throughout the land points - e.g. `/scratch/v14/pas548/restarts/KEEP/restart356/ice/iced.1986-04-01-00000.nc`:
![Screen Shot 2020-10-09 at Fri 9-10 4 29pm](https://user-images.githubusercontent.com/31054815/95546616-c7c06880-0a4c-11eb-82e0-ce07da605658.png)

However I haven't looked at the other restart fields or files so maybe there would be problems with them?

I guess a safe thing would be to set the _FillValue for each field to the value from a land point that isn't masked out?

**Comment by aekiss** [comment]
For example the point (474, 2613) is land but unmasked so you could check its value for every field in every restart file in `/scratch/v14/pas548/restarts/KEEP/restart356/ice/` and use this as the _FillValue

**Comment by aidanheerdegen** [comment]
CF conventions allow for `_FillValue` and `missing_value`. Is `missing_value` is set to something that is non-zero does that help?

**Comment by aekiss** [comment]
Thanks @nichannah, I'm closing this issue now.

We decided in the 14 Oct TWG meeting that this issue with restarts is not significant enough to warrant fixing, and that a fix with a change to `_FillValue`=0 would cause more trouble than it was worth, since genuine data could be misinterpreted as fill.

We just need to remember to fill in the cpu-masked cells with zero values if restarting with a changed cpu layout.

I've done a test run at 0.1deg with PIO ([using commit 7c74942](https://github.com/COSIMA/cice5/tree/7c749428478abde0c2183477168266e90cbb1f1c)) to compare to one without PIO ([using commit 26e6159](https://github.com/COSIMA/cice5/tree/26e6159)).
This restart issue means I can't compare the restart files, but I've confirmed (using xarray's `identical` method) that the outputs are identical, including for a second run based on PIO-generated restarts, so I'm confident that the model state is unaffected by these differences in the restart files. Test script is here: https://github.com/aekiss/notebooks/blob/72986342795e6fef167ad5d9df76a01b1ad7fefa/check_pio.ipynb

**Comment by aekiss** [comment]
Sorry @nic, I'm reopening again - I've hit a bug using PIO in a 1deg configuration.

For the 1deg config I'm using one core per chunk, laid out the same way as `slenderX1` (not sure if this is the best choice?)
```
    history_chunksize_x = 15
    history_chunksize_y = 300
```

I have repeated identical runs 
```
/home/156/aek156/payu/testing/all-configs/v2.0.0rc9/1deg_jra55_ryf_v2.0.0rc9
/home/156/aek156/payu/testing/all-configs/v2.0.0rc9/1deg_jra55_ryf_v2.0.0rc9xx
```
and got differing output in these files and variables:
```
/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9-CHUCKABLE/output000/ice/OUTPUT/iceh.1900-01.nc
/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9xx-CHUCKABLE/output000/ice/OUTPUT/iceh.1900-01.nc
fsurfn_ai_m
vicen_m

/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9-CHUCKABLE/output000/ice/OUTPUT/iceh.1900-02.nc
/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9xx-CHUCKABLE/output000/ice/OUTPUT/iceh.1900-02.nc
fmelttn_ai_m
vicen_m

/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9-CHUCKABLE/output001/ice/OUTPUT/iceh.1900-04.nc
/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9xx-CHUCKABLE/output001/ice/OUTPUT/iceh.1900-04.nc
aicen_m
flatn_ai_m
fmelttn_ai_m

/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9-CHUCKABLE/output001/ice/OUTPUT/iceh.1900-05.nc
/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9xx-CHUCKABLE/output001/ice/OUTPUT/iceh.1900-05.nc
flatn_ai_m
fsurfn_ai_m
vicen_m

/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9-CHUCKABLE/output002/ice/OUTPUT/iceh.1900-07.nc
/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9xx-CHUCKABLE/output002/ice/OUTPUT/iceh.1900-07.nc
fcondtopn_ai_m
```
Note that this issue only appears in multi-category variables (e.g. `'fcondtopn_ai_m' (time: 1, nc: 5, nj: 300, ni: 360)`) and is unpredictable - most multi-category variables are ok most of the time, and there are no variables that are always affected.

For example here's category 0 of `fmelttn_ai_m` in 
`/scratch/x77/aek156/1deg_jra55_ryf_v2.0.0rc9xx-CHUCKABLE/output000/ice/OUTPUT/iceh.1900-02.nc`
![Screen Shot 2020-10-28 at Wed 28-10 9 49am](https://user-images.githubusercontent.com/31054815/97370851-0b97e680-1904-11eb-8f1d-1f7abdc5bddc.png)
There are bad points just north of the Equator over a limited longitude range in the Indonesian archipelago. They are extremely large, presumably uninitialised values. The values in the longitudes between them are very small but nonzero (they should be zero). The land mask is also messed up.

The problem occurs in different places in other fields.

I've only seen this problem in category 0, but I haven't checked thoroughly.
e.g here's category 1 of the same field and file:
![Screen Shot 2020-10-28 at Wed 28-10 9 50am](https://user-images.githubusercontent.com/31054815/97370858-0fc40400-1904-11eb-8164-129538904bdc.png)

I didn't see this issue with the 0.1deg config. Maybe I need better choices for `history_chunksize_x` and `history_chunksize_y`? (NB I found I could get segfaults if I wasn't careful with these values...)

**Comment by aekiss** [comment]
Oops, apologies @nichannah - this was just because I was calling `mpirun` with the wrong options at 1 deg.

When I use
```
mpirun: --mca io ompio --mca io_ompio_num_aggregators 1
```
in `config.yaml` it works as expected.

**Comment by aidanheerdegen** [comment]
The OpenMPI docs say `ompio` is the default for versions > `2.x`. Is that incorrect?

https://www.open-mpi.org/faq/?category=ompio

**Comment by nichannah** [comment]
On Gadi it appears that romio is used by default. Also we need to specify the number of MPI aggregators explicitly to avoid the heuristic/algorithm that usually sets this. This algorithm appears to get confused with the combination of (chunksize != tile size) and deflation on. The confusion leads to a divide-by-zero. I haven't spent the time to really understand this bug/problem so you could say that `--mca io_ompio_num_aggregators 1` is a work-around.

**Comment by aidanheerdegen** [comment]
Thanks for the explanation @nichannah

**Comment by aekiss** [comment]
@nic FYI: PIO seems to slow down CICE at 1 deg.
see 3-month runs in `/home/156/aek156/payu/testing/all-configs/v2.0.0rc9`

Fraction of MOM runtime in oasis_recv, Max CICE I/O time (s)
1 deg, no PIO `1deg_jra55_ryf_v2.0.0rc9_nopio`: 0.04, 10.6
1 deg, PIO with 24 chunks (15x300) `1deg_jra55_ryf_v2.0.0rc9_pio`: 0.07, 17.1
1 deg, PIO with 1 chunk (360x300) `1deg_jra55_ryf_v2.0.0rc9_pio_1chunk`: 0.096, 24.4

but it is improved at 0.25deg:
0.25 deg, no PIO `025deg_jra55_ryf_v2.0.0rc9`: 0.078, 54
0.25 deg, PIO with 100 chunks (144x108) `025deg_jra55_ryf_v2.0.0rc9_pio2`: 0.04, 25

The cice cores are spread between nodes on gadi at 1deg with 1+216+24 cores for yatm/mom/cice so that might be part of the problem: https://github.com/COSIMA/access-om2/issues/212 and https://github.com/COSIMA/access-om2/issues/202

**Comment by aekiss** [comment]
I've also tried 1 deg (`/home/156/aek156/payu/testing/all-configs/v2.0.0rc10/1deg_jra55_iaf_v2.0.0rc10`) and 0.25 deg (`025deg_jra55_iaf_v2.0.0rc10`) configs with 4 chunks (90x300 at 1 deg; 720x540 at 0.25 deg) and get 0.085 for the fraction of MOM runtime in oasis_recv in both cases.

1 deg with 4 chunks is almost as fast as the 24-chunk case (though slower than without PIO) but should be faster to read in most circumstances than 24 chunks. However I'm thinking a 180x150 4-chunk layout is probably a better match to hemisphere-based access patterns so I might try that too.

0.25 deg with 4 chunks is now somewhat slower than without PIO but I'm reluctant to use too many chunks in case it slows down reading.

Also I should have mentioned that these 1 deg and 0.25 deg tests all had identical ice outputs, but they differ from the ice outputs in the production 0.1deg runs I reported [here](https://github.com/COSIMA/cice5/issues/34#issuecomment-687435055) so they aren't directly comparable to those.

**Comment by aekiss** [comment]
Some more tests of differing `history_chunksize_x` x `history_chunksize_y` with 3mo runs at 1 deg in `/home/156/aek156/payu/testing/all-configs/v2.0.0rc`:

Fraction of MOM runtime in oasis_recv, Max CICE I/O time (s)
1 deg, PIO with 4 chunks (90x300) `1deg_jra55_iaf_v2.0.0rc10_3mo`: 0.067, 16.8
1 deg, PIO with 4 chunks (180x150) `1deg_jra55_iaf_v2.0.0rc10_3mo_180x150`: 0.072, 18.1

The first of these is slightly faster (presumably because it is consistent with the 15x300 core layout) but the difference is small and so I will use 180x150 for the new 1deg configs as this is better suited to typical access patterns of reading one hemisphere or the other.

The fraction of MOM runtime in oasis_recv values with 180x150 is smaller in the 3 mo case compared to 5yr: 0.072 rather than 0.085 (see prev post). So for 3mo runs the 4-chunk cases are essentially as fast as the 24-chunk case.

**Comment by aekiss** [comment]
For future reference: the [processor masking in the ice restarts](https://github.com/COSIMA/cice5/issues/34#issuecomment-696491335) can be fixed with https://github.com/COSIMA/topogtools/blob/master/fix_ice_restarts.py, allowing a change in processor layout during a run.
